### PR TITLE
@erikdstock => dont make users enter credit card again when they try to bid and already have one on file

### PR DESCRIPTION
**Pull request by sweir27**
This PR addresses: https://artsyproduct.atlassian.net/browse/PURCHASE-85

We now return whether or not a user has a valid credit card through the `has_valid_credit_cards` field under `me` (from metaphysics).

I updated the flow slightly to also grab whether or not a user is registered for the sale in the same metaphysics query as above. It's not really any more efficient but it's a little easier to follow this logic than to have to poke through backbone models.

For clarity, here's what the logic does:
- when you click "Bid" on the artwork page, we check to see if you are `isRegistered` and `hasValidCreditCard`.
  - If `isRegistered` is true, then we show you a confirm bid page with no cc form attached
    - when you click "bid" we just create a bidder position via the `me/bidder_positions` endpoint
  - If `isRegistered` is false but `hasValidCreditCard` is true, we show you a confirm bid page with no cc form attached
    - when you click "bid", we also _only_ call the `me/bidder_positions` endpoint which [has logic](https://github.com/artsy/gravity/blob/master/app/api/v1/me_bidder_positions_endpoint.rb#L61-L66) to create a bidder before creating the bidder position
  - If `isRegistered` is false and `hasValidCreditCard` is false, we show you a confirm bid page with a cc form attached
    - when you submit the form, we first call `/bidder` to create a bidder, and then call `me/bidder_positions` to create the position

Leaving this as WIP because as part of my local QA I noticed that for the case where we just call `me/bidder_positions` to both create the bidder _and_ the bidder position, we aren't triggering an [email from pulse](https://github.com/artsy/gravity/blob/master/app/api/v1/bidders_endpoint.rb#L67) as we do from the `/bidder` endpoint. I'll follow up with a gravity PR to do that.

(blocking) gravity PR: https://github.com/artsy/gravity/pull/11702

### CODE REVIEW

File: src/desktop/apps/auction_support/routes.coffee [hunk 1]
```diff
@@ -109,25 +110,26 @@ registerOrRender = (sale, req, res, next) ->
...
-  # TODO: Refactor all of this junk to use MP, or drop it in favor of
-  # inline bidding component, see: https://github.com/artsy/force/issues/5118
   metaphysics
     query: """ {
       artwork(id: "#{req.params.artwork}") {
         sale_artwork {
           bid_increments
         }
       }
+      me {
```
Comment (sweir27) [inline src/desktop/apps/auction_support/routes.coffee]: Combined this into one query

**Comment by sweir27** [comment]
Verified locally and the gravity PR is deployed so this is also ready for 👀

File: src/desktop/apps/auction_support/routes.coffee [hunk 2]
```diff
@@ -79,7 +79,7 @@ registerOrRender = (sale, req, res, next) ->
   saleArtwork = new SaleArtwork(artwork: new Artwork(id: req.params.artwork), sale: sale)
   bidderPositions = new BidderPositions(null, { saleArtwork: saleArtwork, sale: sale })
 
-  render = _.after 4, ->
+  render = _.after 3, ->
     res.locals.sd.BIDDER_POSITIONS = bidderPositions.toJSON()
     res.locals.sd.SALE = sale.toJSON()
```
Comment (erikdstock) [inline src/desktop/apps/auction_support/routes.coffee]: Why the change from 3 - 4?

Comment (sweir27) [reply src/desktop/apps/auction_support/routes.coffee]: Because I removed the `checkRegisteredForAuction` query below (since it got combined into the metaphysics one).

File: src/desktop/apps/auction_support/routes.coffee [hunk 3]
```diff
@@ -109,25 +110,26 @@ registerOrRender = (sale, req, res, next) ->
...
+        bidders(sale_id: "#{sale.get('id')}") {
+          id
+        }
+      }
     } """
-  .catch(next).then ({ artwork }) ->
+    req: req
+  .catch(next).then ({ artwork, me }) ->
     res.locals.bidIncrements = artwork.sale_artwork.bid_increments
+    res.locals.sd.REGISTERED = Boolean(me && me.bidders && me.bidders.length > 0)
```
Comment (mzikherman) [inline src/desktop/apps/auction_support/routes.coffee]: maybe `IS_REGISTERED` is better?

also ⛳️ its ☕️ so can do `me?.bidders?.length?` I believe

File: src/desktop/apps/auction_support/routes.coffee [hunk 4]
```diff
@@ -109,25 +110,26 @@ registerOrRender = (sale, req, res, next) ->
...
+    res.locals.sd.HAS_VALID_CREDIT_CARD = Boolean(me && me.has_qualified_credit_cards)
```
Comment (mzikherman) [inline src/desktop/apps/auction_support/routes.coffee]: `me?.has_qualified_credit_cards`

File: src/desktop/apps/auction_support/routes.coffee [hunk 5]
```diff
@@ -109,25 +110,26 @@ registerOrRender = (sale, req, res, next) ->
...
   metaphysics
     query: """ {
       artwork(id: "#{req.params.artwork}") {
         sale_artwork {
           bid_increments
         }
       }
+      me {
+        has_qualified_credit_cards
+        bidders(sale_id: "#{sale.get('id')}") {
```
Comment (mzikherman) [inline src/desktop/apps/auction_support/routes.coffee]: probably better to structure this as a proper query w/ vars passed in, rather than interpolate (now that there's more than one var being passed in).

so something like:

```
query($sale_id: String!, $artwork_id: String!) {
  artwork(id: $artwork_id) {

   }
  me {
    bidders(sale_id: $sale_id) {

    }
  }
}
```

File: src/desktop/apps/auction_support/routes.coffee (hunk 3 above)
Comment (sweir27) [reply src/desktop/apps/auction_support/routes.coffee]: gah ☕️
